src/v/coproc/wasm_event_listener.cc

```diff
@@ -0,0 +1,219 @@
+          "loop, after it has already been started or closed");
+    }
+    return storage::directories::initialize(_active_dir.string())
+      .then([this] {
+          return storage::directories::initialize(_inactive_dir.string());
+      })
+      .then([this] {
+          return storage::directories::initialize(_submit_dir.string());
+      })
+      .then([this] {
+          (void)ss::with_gate(_gate, [this] {
+              return _client.connect().then([this] {
+                  return ss::do_until(
+                    [this] { return _abort_source.abort_requested(); },
+                    [this] { return do_start(); });
+              });
+          });
+      });
+}
+
+ss::future<> wasm_event_listener::do_start() {
+    /// This method performs the main polling behavior. Within a repeat loop it
+    /// will poll from data until it cannot poll anymore. Normally we would be
+    /// concerned about keeping all of this data in memory, however the topic is
+    /// compacted, we don't expect the size of unique records to be very big.
+    return ss::do_with(
+      model::record_batch_reader::data_t(),
+      [this](model::record_batch_reader::data_t& events) {
+          /// The stop condition is met when its been detected that the stored
+          /// offset has not moved.
+          return ss::repeat([this, &events] {
+                     model::offset initial_offset = _offset;
+                     return poll_topic(events).then([this, initial_offset] {
+                         return initial_offset == _offset
+                                  ? ss::stop_iteration::yes
+                                  : ss::stop_iteration::no;
+                     });
+                 })
+            .then([this, &events] { return reconcile_wasm_events(events); })
+            .then([this](wasm_script_actions wsas) {
+                return ss::do_with(
+                  std::move(wsas), [this](wasm_script_actions& wsas) {
+                      return ss::do_for_each(
+                        wsas, [this](wasm_script_actions::value_type& wsa) {
+                            /// Finally the action is performed, either the
+                            /// source is written to disk or a file is deleted
+                            /// from disk
+                            return resolve_wasm_script(
+                              wsa.first, std::move(wsa.second));
+                        });
+                  });
+            })
+            .then([this] { return ss::sleep_abortable(2s, _abort_source); })
+            .handle_exception([](std::exception_ptr eptr) {
+                vlog(coproclog.debug, "Exited sleep early: {}", eptr);
+            });
+      });
+}
+
+/// Map a model::record_batch -> std::vector<wasm_script_action>
+/// The transform will only occur after the record has passed all validators
+wasm_event_listener::wasm_script_actions
+wasm_event_listener::reconcile_wasm_events(
```

Comment:
This walks and talks like a `record_batch_reader`, but looks different.

Check out `batch_validating_consumer` in one of the tests.

Reply:
Oh, the values are shared, not consumed. Hmm, so close, though.